src/Microsoft.AspNetCore.Sockets.Client.Http/WebSocketsTransport.cs

```diff
@@ -132,77 +172,120 @@ public WebSocketsTransport(HttpOptions httpOptions, ILoggerFactory loggerFactory
 
                     if (receiveResult.EndOfMessage)
                     {
-                        await _application.Output.FlushAsync(_transportCts.Token);
+                        var flushResult = await _application.Output.FlushAsync();
+
+                        // We cancelled in the middle of applying back pressure
+                        // or if the consumer is done
+                        if (flushResult.IsCancelled || flushResult.IsCompleted)
+                        {
+                            break;
+                        }
                     }
                 }
             }
             catch (OperationCanceledException)
             {
                 _logger.ReceiveCanceled();
             }
+            catch (Exception ex)
+            {
+                if (!_aborted)
+                {
+                    _application.Output.Complete(ex);
+
+                    // We re-throw here so we can communicate that there was an error when sending
+                    // the close frame
+                    throw;
+                }
+            }
             finally
             {
                 // We're done writing
-                _logger.ReceiveStopped();
-                _transportCts.Cancel();
+                _application.Output.Complete();
             }
+
+            _logger.ReceiveStopped();
         }
 
-        private async Task SendMessages()
+        private async Task StartSending(WebSocket ws)
         {
-            _logger.SendStarted();
-
             var webSocketMessageType =
                 Mode == TransferMode.Binary
                     ? WebSocketMessageType.Binary
                     : WebSocketMessageType.Text;
 
+            Exception error = null;
+
             try
             {
                 while (true)
                 {
-                    var result = await _application.Input.ReadAsync(_transportCts.Token);
+                    var result = await _application.Input.ReadAsync();
                     var buffer = result.Buffer;
+
+                    // Get a frame from the application
+
                     try
                     {
-                        if (!buffer.IsEmpty)
+                        if (result.IsCancelled)
                         {
-                            _logger.ReceivedFromApp(buffer.Length);
+                            return;
```

Comment:
Use `break` as above to ensure we hit the logging below? Or use `return` above. Just be consistent ;)